src/js/integrations/stalltracking.ts

```diff
@@ -0,0 +1,344 @@
+/* eslint-disable max-lines */
+import { IdleTransaction, Span, Transaction } from "@sentry/tracing";
+import { Integration, Measurements } from "@sentry/types";
+import { logger, timestampInSeconds } from "@sentry/utils";
+
+export interface StallMeasurements extends Measurements {
+  stall_count: { value: number };
+  stall_total_time: { value: number };
+  stall_longest_time: { value: number };
+}
+
+export type StallTrackingOptions = {
+  /**
+   * How long in milliseconds an event loop iteration can be delayed for before being considered a "stall."
+   * @default 100
+   */
+  minimumStallThreshold: number;
+};
+
+/** Margin of error of 20ms */
+const MARGIN_OF_ERROR_SECONDS = 0.02;
+/** How long between each iteration in the event loop tracker timeout */
+const LOOP_TIMEOUT_INTERVAL_MS = 50;
+
+/**
+ * Stall measurement tracker inspired by the `JSEventLoopWatchdog` used internally in React Native:
+ * https://github.com/facebook/react-native/blob/006f5afe120c290a37cf6ff896748fbc062bf7ed/Libraries/Interaction/JSEventLoopWatchdog.js
+ *
+ * However, we modified the interval implementation to instead have a fixed loop timeout interval of `LOOP_TIMEOUT_INTERVAL_MS`.
+ * We then would consider that iteration a stall when the total time for that interval to run is greater than `LOOP_TIMEOUT_INTERVAL_MS + minimumStallThreshold`
+ */
+export class StallTracking implements Integration {
+  /**
+   * @inheritDoc
+   */
+  public static id: string = "StallTracking";
+
+  /**
+   * @inheritDoc
+   */
+  public name: string = StallTracking.id;
+
+  public isTracking: boolean = false;
+
+  private _minimumStallThreshold: number;
+
+  /** Total amount of time of all stalls that occurred during the current tracking session */
+  private _totalStallTime: number = 0;
+  /** Total number of stalls that occurred during the current tracking session */
+  private _stallCount: number = 0;
+
+  /** The last timestamp the iteration ran in milliseconds */
+  private _lastIntervalMs: number = 0;
+  private _timeout: ReturnType<typeof setTimeout> | null = null;
+
+  private _longestStallsByTransaction: Map<Transaction, number> = new Map();
```

Comment:
Can we use `WeakMap`? We don't need to have a strong reference here and it could leak if for some reason `finish` isn't called.
This suggestion applies to any map of <Transaction, X>

Reply:
Replaced with a max running transaction count as discussed due to WeakMap not allowing us to iterate through the keys or with a way to know how many transactions we are tracking.